src/service/resources-impl.js

```diff
@@ -407,13 +410,29 @@ export class Resources {
    * @private
    */
   buildReadyResources_() {
+    // Avoid cases where elements add more elements inside of them
+    // and cause an infinite loop of building - see #3354 for details.
+    if (this.isCurrentlyBuildingPendingResources_) {
+      return;
+    }
+    try {
+      this.isCurrentlyBuildingPendingResources_ = true;
+      this.buildReadyResourcesUnsafe_();
+    } finally {
+      this.isCurrentlyBuildingPendingResources_ = false;
+    }
+  }
+
+  /** @private */
+  buildReadyResourcesUnsafe_() {
     for (let i = 0; i < this.pendingBuildResources_.length; i++) {
       const resource = this.pendingBuildResources_[i];
-      if (this.documentReady_ || hasNextNodeInDocumentOrder(resource.element)) {
-        resource.build();
-        // Resource is built remove it from the pending list and step back
-        // one in the index to account for the removed item.
+      if (this.documentReady_ ||
+          hasNextNodeInDocumentOrder(resource.element)) {
+        // Remove resource before build to remove it from the pending list
+        // in either case the build succeeed or throws an error.
```

Comment:
Yeah you're right. I kinda skipped this `resource.build` method and was looking at the `element.build`.

This might just to be safe not to get stuck in that state of deadlocking. But @cramforce what do you think?